---
title: "Genome Function Phylogenetics"
author: "Dennis Psaroudakis"
date: "April 28th 2019"
keep_md: yes
output:
  tufte::tufte_handout:
    latex_engine: xelatex
    citation_package: natbib
  tufte::tufte_html: default
bibliography: resources/bib.bib
link-citations: yes
classoption: justified
language: en-US
header-includes:
  - \usepackage[osf]{libertine}
  - \citestyle{plain}
---

```{r setup, include=FALSE}
library(tufte)
# invalidate cache when the tufte version changes
knitr::opts_chunk$set(tidy = FALSE, cache.extra = packageVersion('tufte'))
options(htmltools.dir.version = FALSE)
```

`r newthought("When we build phylogenetic trees,")` we often use the DNA or protein sequences of certain genes^[with appropriate degree of conservedness depending on the evolutionary distance of the taxa, little probability of horizontal gene transfer, reliable for sequencing and alignment etc.] as the basis for our analysis.
They are the direct substance of evolution so that makes sense, especially when we need high resolution (i.e. a signal in very closely related organisms).
In some cases, they don't give the answer we're looking for though:
Let's say for example, that you're a doctor and one of your patients is infected with a novel pathogen.
You can sequence that pathogen and place it somewhere in the bacterial taxonomy, and naturally you will start treating the patient based on what you know about closely related species.
Unfortunately, none of the antibiotics you're using show any effect.
How come?
Just because two organisms are closely related does not *necessarily* mean that they are similar in their phenotype.
Especially if you've only used genes that are suitable for regular phylogenetics but don't actually have anything to do with the organism's metabolism, membrane structure etc, you have no guarantee that close evolutionary relationship means similar response to environmental factors.
In this case, a tree that can group your pathogen with organisms based on their actual *in vivo* similarity would be much more helpful here.
There are trees built on phenotypic characteristics (nose color, ear length...) but the choice of what characteristics to look at is not trivial and has an incredible influence on the result.
You will never be able to cover all relevant attributes and your choice can therefore be considered somewhat arbitrary\sidenote{
Look at this (parodic) taxonomy of animals by Jorge Luis Borges:
1. those that belong to the Emperor,
2. embalmed ones,
3. those that are trained,
4. suckling pigs,
5. mermaids,
6. fabulous ones,
7. stray dogs,
8. those included in the present classification,
9. those that tremble as if they were mad,
10. innumerable ones,
11. those drawn with a very fine camelhair brush,
12. others,
13. those that have just broken a flower vase,
14. those that from a long way off look like flies.\medskip
}.
A method would be needed that is exhaustive, clearly defined, and reproducible, but also more effective at finding the answer you're looking for than using the DNA sequence.
And, unsurprisingly, there is [@Zhu2015] (it's a cool paper, check it out)!

In this paper, I'm going to do something less exciting. Instead of working with life-threatening bacteria and saving patients, I will focus on plants.
The idea is somewhat similar though^[and I had the idea myself first and only found the paper afterwards]: I want to build a phylogenetic tree of plant species, not based on their genetic sequence but on the processes, structures, and reactions that are present in it.

# The Gene Ontology
Historically, the function/role that a gene plays in an organism has been described in natural language, however the researcher characterizing that gene deemed best.
While this is nice to read, it is not very useful if you want to do computation on it, as computers are (still) horrible at understanding natural language and determining the structure in meaning behind the words.
Additionally, different people will describe the same thing with different words, so there is always a potential for misunderstanding.

Ontologies try to alleviate these problems by providing a strictly organized and controlled vocabulary and defined relationships between the terms, so that the same statement always means the same thing, no matter the context or the author. Additionally, ontologies can be understood by computers if all relationships and terms are clearly defined.

The Gene Ontology (GO) is such an ontology.
It describes genes by the properties of their product.
\marginnote{"The mission of the GO Consortium is to develop an up-to-date, comprehensive, computational model of biological systems, from the molecular level to larger pathways, cellular and organism-level systems."

\raggedleft --- GO Consortium (geneontology.org)}
In our case these gene products are proteins, and they can be characterized in three different aspects:

- What biological processes is this protein part of? (e.g. photosynthesis or autophagy)
- What molecular functions does the protein carry out? (e.g. ethylene binding or RNA ligation)
- What cellular component is the protein active at? (e.g. outer membrane or nucleus)

Within each of these aspects, the Gene Ontology defines a huge number of terms (2,675,070 in total), that range from very general to very specific:


![subtree of the *Biological Process* ontology. 
The terms are organized in such a way that more general terms are always true for any of their more specific child terms.
For example, any protein that is part of *Cytokinesis after Meiosis I*, is also obviously part of Cytokinesis, the Cell Cycle etc.
That way, a gene that has been annotated with the term *Cytokinesis after Meisosis I* (leaf term), has implicitly been annotated with all of that term's parent terms as well, all the way up to the root term.](resources/go_hierarchy_example.png){width=60%}

\noindent When proteins are annotated with these terms instead of just natural language, we can now computationally answer some interesting questions, such as:

- How similar in function^[*Function* refers to any aspect of the GO, not just Molecular Function.] is protein A to protein B? (One answer would be: How many steps in the GO graph do I need from term A to B? The fewer steps, the more similar the function)
- If protein A is involved in Biological Process XYZ, what other proteins are involved in that same process?

\noindent The Gene Ontology is quite well established in the field, so you will find GO annotations all across UniProt entries or be able to use dedicated tools like [AmiGO](http://amigo.geneontology.org/amigo) or [QuickGO](https://www.ebi.ac.uk/QuickGO/) to examine a protein of interest.

# Data
Annotating genes with their functions can be done experimentally (e.g. by knocking out a certain gene and seeing what processes in the cell are affected), but that is a time-consuming and expensive process, so methods have been developed that try to predict the function of a given gene.
Our lab has developed such a pipeline called [Gene Ontology Meta Annotator for Plants (GOMAP)](https://dill-picl.org/projects/gomap/) which combines different prediction approaches and is able to generate high-confidence and very extensive Functional Annotations in a reproducible manner [@Wimalanathan2018].
We have been applying this pipeline to whole-genome assemblies of different plant species and generated functional annotations for every gene in each genome.
These annotation sets are (or will be shortly) available from  
\scriptsize https://dill-picl.org/projects/gomap/gomap-datasets \normalsize.

# Method
I am using two different tree building approaches, one is a distanced based method and the other a parsimony one.

## Distance Based
Starting point of our analysis are the functional annotation sets, one for each genome, which annotate every gene in the genome with one or more GO terms.
In more mathematical terms the genome annotation set is a list of tuples $(G, T)$ with $G \in$ Genes in that genome and $T \in$ Terms in the Gene Ontology.
\marginnote[-5em]{
\texttt{Gene\ \ \ \ \ \ \ \ \ GO Term\\
Os01g0601625	GO:0050896\\
Os01g0601625	GO:0016021\\
Os01g0601625	GO:0016301\\
Os01g0601651	GO:0003677\\
Os01g0601651	GO:0009699\\
Os01g0601651	GO:0050790\\
Os01g0601651	GO:0050794\\
Os01g0601651	GO:0050896\\
Os01g0601675	GO:0007275\\
Os01g0601675	GO:0016310\\
Os01g0601675	GO:0050789\\
...
}
}

We can use the hierarchical structure of the Gene Ontology to obtain the ancestors $A_i$ of any term $T_i$; in other words the gene $G_i$ is not just annotated with the term $T_i$ itself but also with all GO terms that are a more general statement of that term (e.g. any gene that is part of a metabolic process is thereby also part of a biological process).
We do that for all terms $T$ in the dataset and combine all of the terms and their ancestors into one big genome-wide set $S$, irrespectively of the gene they were originally associated with:  $S = \bigcup_{i=1}^{x}(T_i \cup A_{i})$.

When this superset of annotations is created for each of the datasets, we can use the Jaccard Distance as a measure of how (dis-)similar any two sets are from each other, or in biological terms how different the two genomes are on a functional level:

$$ \textrm{Jaccard Distance}(S_a, S_b) = 1 - \frac{|S_a \cap S_b|}{|S_a \cup S_b|} $$

\begin{marginfigure}
The general idea of using the Jaccard Distance in this context is to measure the overlap of two subtrees in the GO hierarchy.
Say, for simplicity, that we're looking at two genomes (here called Gold Standard and Prediction) that each only contain one single GO term (marked by a leaf).
First, we add all ancestors of that leaf term to each subtree.
Then, we determine the overlap (which corresponds to $S_a \cap S_b$), and divide the number of nodes in this overlap by the number of nodes in either of the two subtrees ($S_a \cup S_b$).
\vspace{0.2em}

\includegraphics{resources/hF-graphical.png}

\vspace{0.5em}
\noindent In the case of this example, the Jaccard Distance of Gold Standard and Prediction would be $1 - \frac{4}{9} = \frac{5}{9}$
\end{marginfigure}

Applying this formula to all pairwise combinations of the genomes we're looking at yields a $S \times S$ distance matrix that can then serve as the input for a neighbor joining algorithm (provided by PHYLIP).
I rooted the resulting tree manually outside of the grasses (maize, wheat, rice, barley).

## Parsimony Based

Like in the previous method, we again start by enriching all taxa sets $(G,T)$ with their ancestor terms and discarding the gene association: $S = \bigcup_{i=1}^{x}(T_i \cup A_{i})$.

Instead of using these sets for distance calculation, we combine them into a big binary matrix that displays which terms are present in which set:

\begin{table}[]
\begin{tabular}{lllllll}
Taxon       & \scriptsize{GO:0016021} & \scriptsize{GO:0009699} & \scriptsize{GO:0050794} & \scriptsize{GO:0050789} & \scriptsize{GO:0060739} & ... \\
\textit{G. max}      & 1          & 0          & 1          & 0          & 1          &     \\
\textit{T. aestivum} & 1          & 1          & 0          & 1          & 1          &     \\
...         &            &            &            &            &            &    
\end{tabular}
\end{table}

This matrix was then used as the input for `pars` from the PHYLIP package to find the maximally parsimonious tree.

# Results

The phylogram created by the distance based method is displayed in Figure \ref{fig:result}, the maximum parsimony tree in figure \ref{fig:parsimony}.

\begin{figure*}[htbp]
\includegraphics{resources/result.png}
\caption{Phylogram built on the Jaccard distance matrix with Neighbor Joining. Manually rooted between monocots and dicots and text in green added.}
\label{fig:result}
\end{figure*}

\begin{figure*}[htbp]
\includegraphics{resources/parsimonyTree.png}
\caption{Phylogram built by looking for the maximally parsimonious tree (total of 7780 changes). Manually rooted between monocots and dicots.}
\label{fig:parsimony}
\end{figure*}

Both trees were combined into a tanglegram (see figure \ref{fig:tanglegram}).

\begin{figure*}[htbp]
\includegraphics{resources/Tanglegram.png}
\caption{Tanglegram of the distance based and maximum parsimony tree (both manually rooted).}
\label{fig:tanglegram}
\end{figure*}

# Discussion

## Fulfilling expectancies?
In a perfect world, the evolution of organisms could be just as easily retraced by the evolution of their functions as by their sequence; after all the selective pressure is on the mechinisms and activities of an organism and not on its DNA sequence and what differentiates two species from each other are the differences in their *in vivo* pheontype. \marginnote{Although you may get lower resolution, because silent mutations or such that don't alter the function of a protein would be missed.}

The expected taxonimic tree is displayed in figure \ref{fig:expectedTree} and indeed the tree produced by the distance based method is quite similar (much more similar than I had expected).
There are only two notable differences: *Sorghum* should be at the base of maize and not grouped with *Brachypodium* and *Medicago* and *Arachis* should switch their places.
We are currently investigating the reason for this phenomenon and the best explanation we've come up with so far is that *Brachy* and *Medicago* are not actually good representations of their respective group; they have been chosen as model organisms because they are easy to sequence but that's mainly because their genome is much smaller and less complex than that of actual crop plants. *Medicago* is also well known for having an unusally high rate of genomic evolution, gene births and gene deaths [@Young2011].

\begin{marginfigure}
\includegraphics{resources/expectedTree.png}
\caption{Expected cladogram. \citep{Kellogg2001,Azani2017,Puttick2018}}
\label{fig:expectedTree}
\medskip
\end{marginfigure}

It is difficult to come up with a clear evolutionary tree within the same species (I am talking about maize^[Many of the maize lines have more than one ancestor and there is an immense amount of horizontal gene transfer with maize, even to closely related species.]) but it is encouraging to see that in the generated tree the 3 different classes of maize (stiff stalk, non-stiff stalk, and iodent) are differentiated from each other in a reasonable way.

`r newthought("What about the parsimony tree?")` Surprisingly, the tree built on the parsimony method is not identical with the distance based one, and it is even less similar to the expected tree.
Our first thought was that since the parsimony method does not search the complete tree space, it might simply not have found the distance based tree even though it is the most parsimonious one.
Unfortunately that does not seem to be the case: The tree it found has a score of 7780, the distance based tree 7821. When manually switching *Medicago* and *Arachis* to get closer to the taxonomic tree, the score improved to 7812, but putting *Sorghum* at the place it was expected increased the score again to 7866.
So it seems the answer is not that simple and we will have to reflect more on our methods to understand where the difference comes from and which is the more meaningful result.

## Tracing Back the Signal
Another big question is why the tree actually looks the way it does. Is it truly a biologically meaningful display of differences and similarities in function or is it just an artifactual tree, caused by a bias in the method.
Given the number of taxa it seems unlikely that the similarity to the expected tree is just random noise, so there must be some systematic reason behind it.
To make a first step answering this question, I investigated what actually causes species to cluster the way they do. I asked the following question: 
Which terms are common to all nitrogen fixing plants but do not occur in any other plant in the tree?^[in mathy words this is the intersection of terms in the nitrogen fixing species minus the union of terms in all other species]
If these terms are actually meaningful to the process of nitrogen fixation (or some other characteristic that clearly differentiates this group of plants from the others) that would indicate that the phylogenetic signal actually comes from biologically plausible differences in the functional annotations.
So, here is the answer:

- GO:0080184 - response to phenylpropanoid
- GO:0033800 - isoflavone 7-O-methyltransferase activity
- GO:0042577 - lipid phosphatase activity
- GO:0031174 - lifelong otolith mineralization
- GO:0045299 - otolith mineralization
- GO:0006742 - NADP catabolic process
- GO:0019364 - pyridine nucleotide catabolic process
- GO:0019677 - NAD catabolic process
- GO:0070823 - HDA1 complex

Three of these terms do seem plausible: The Nod Factor molecule that is secreted by the plant as a signal to Rhizobia bacteria is a phenylpropanoid, isoflavone 7-O-methyltransferase activity is involved in Nod Factor synthesis, and the nodule has a complex lipid membrane system – and Nod Factor has a lipid component (which could explain lipid phosphatase activity).
The last 4 terms are possibly plausible, but less certain, but the two otolith terms seem very out of place, since they describe small oval calcareous bodies in the inner ear of vertebrates, not something expected to be found in legumes.
There might still be some biological significance to it but we're not sure yet what that might be. So for now it seems that some of the signal clearly comes from biological differences while some other part might be due to the method.^[I looked at other, similar constellations as well and the answer was the same (e.g. what differentiates non-stiff stalk maize lines from the others).]

## What's next?

Since this is part of a publication I am working on, a lot more thought will be put into, for example, the following questions:

- How similar is this tree to the taxonomic tree?
- What are the reasons for the differences, what are the reasons for the similarities?
- What biases in the construction process need we be aware of?
- What does the tree depict/how do we interpret it?
- In what way could it be scientifically valuable?
- Are there any other taxa that would be good to have in the tree?
- Should we get rid of *Brachypodium* and *Medicago*?
- Is there any further analysis that would be good to do? (e.g. branch support, looking at each GO aspect separately, a more sophisticated measure of GAF similarity...)
- Is rooting the tree there reasonable?


# Reproducing the Tree
To reproduce this tree, follow these steps on a Linux machine that has PHYLIP available:

```bash
git clone https://github.com/Thyra/EEOB563.git paper_dennis # Clone the repository
cd paper_dennis/final_project # change into the directory

# Distance based method
bin/build_distance_matrix annotation_sets/*tree.json > distance_matrix.phy
module load phylip # (if you're on HPC)
neighbor # (use standard options)

# Parsimony based Method
bin/binary_table annotation_sets/*.tree.json > binary.csv
bin/binary2phylip binary.csv > binary.phy
module load phylip
pars # (again, standard options)

# If you're adding more GAFs
bin/gaf2json <new_gaf>
# Then repeat distance or parsimony as above

# To replace GAF filenames with species names in newick tree
bin/rename_taxa <tree.newick> resources/taxa_name_mapping.csv
```

\noindent If any of the binaries fails without a message that says otherwise, you probably need to install some more libraries. If that doesn't help you may have to recompile them for your processor type (they're written in [Crystal](https://crystal-lang.org/), source code is in `src/`)